free_fleet/src/client/Client.cpp

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (C) 2002 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <chrono>
+#include <thread>
+#include <future>
+#include <iostream>
+#include <unordered_set>
+
+#include <rmf_traffic/Time.hpp>
+
+#include <free_fleet/Console.hpp>
+#include <free_fleet/client/Client.hpp>
+#include "internal_Client.hpp"
+
+namespace free_fleet {
+
+//==============================================================================
+void Client::Implementation::set_callbacks()
+{
+  using namespace std::placeholders;
+  middleware->set_pause_request_callback(
+    std::bind(&Implementation::handle_pause_request, this, _1));
```

Comment:
Even though `middleware` is owned by `Client`, we can't really know how an end user might choose to implement `middleware`, so we can't be sure if this lambda might outlive `this` and end up getting called after `this` has destructed.

For that reason, we should follow the recommendation made [here](https://docs.google.com/presentation/d/1FPUOh85col-KyczN2fpxafwcvUJYT1bnStcrxYS5KFQ/edit#slide=id.ge5f47495e2_0_26) and move all these fields and member functions into a shared data structure, and then capture a `std::shared_ptr` or `std::weak_ptr` of that shared data structure.

Note that the `middleware` field should be left out of the shared data structure to avoid a circular dependency.

Reply:
Moved fields into a separate data structure, and passed shared pointers to lambdas, https://github.com/open-rmf/free_fleet/pull/76/commits/e8bec64aa9dda96a1fe6374b04e1a7258831c076